src/main/kotlin/com/github/imflog/schema/registry/tasks/download/DownloadTaskAction.kt

```diff
@@ -33,6 +34,29 @@ class DownloadTaskAction(
         return errorCount
     }
 
+    private fun expandSubjectPatterns() : List<DownloadSubject> {
+        val allSubjects = client.allSubjects
+        return subjects.flatMap { downloadSubject ->
+            if (downloadSubject.regex) {
+                val regex = parseSubjectRegex(downloadSubject.subject)
+                allSubjects.filter { subject -> regex?.matches(subject) ?: false }
+                    .map { subject -> DownloadSubject(subject, downloadSubject.file, downloadSubject.version) }
+                    .toList()
+            } else {
+                listOf(downloadSubject)
+            }
+        }
+    }
```

Comment:
```suggestion
    private fun expandSubjectPatterns(): List<DownloadSubject> {
        val allSubjects = if (subjects.any { it.regex }) client.allSubjects else listOf()
        return subjects.flatMap { downloadSubject ->
            if (downloadSubject.regex) {
                val regex = parseSubjectRegex(downloadSubject.subject)
                allSubjects
                    .filter { subject -> regex?.matches(subject) ?: false }
                    .map { subject -> DownloadSubject(subject, downloadSubject.file, downloadSubject.version) }
                    .toList()
            } else {
                listOf(downloadSubject)
            }
        }
    }
```

This should avoid calling the subject list when not needed.

Reply:
Yes... did consider doing it at the time but it just feels wasteful because we are searching through all the subjects twice, once to check if there is a regex and then another time to process them. I know given the amount of subjects that there is likely to be that this will probably never be a problem but it annoys me :-)

As an alternative I have merged another suggestion where I use a memorized supplier. The supplier shouldn't be called unless there is a regex and as the result is memorized/cached it will only be called once no matter how many times we fetch the value. On the other hand it also means we don't have to loop through the list of subjects twice.

Reply:
I thought the same about the double run (not efficient but given the amount should be ok).
The memorized version is the best as we ensure that we only call when needed so I think It makes sense to do this.
Good job :)